Custom pupil detection plugins (#396)
pupil-labs/pupil-docs: src/developer/core/plugin-api.md

class MyCustomPlugin(Plugin):
    pass
```

### Pupil Detection Plugins

> Pupil Core supports custom pupil detection plugins which run in the eye process. These plugins can be used to implement custom algorithms for finding the pupil position in the image, and feed them to the rest of the gaze mapping pipeline.

Reviewer: LGTM @papr

> Pupil detection plugins are supported both in Pupil Capture, as well as Pupil Player. Similar to regular user plugins, custom pupil detection plugins should be placed inside `pupil_capture_settings/plugins` for Capture, and `pupil_player_settings/plugins` for Player, both of which are located in the user directory of your system. The plugins are automatically loaded when either of the applications are started.

Reviewer 2: ```suggestion
Pupil detection plugins are supported in Pupil Capture, Pupil Service, and Pupil Player. Similar to regular user plugins, custom pupil detection plugins should be placed inside `pupil_capture_settings/plugins` for Pupil Capture, `pupil_service_settings/plugin` for Pupil Service, and `pupil_player_settings/plugins` for Pupil Player, both of which are located in the user directory of your system. The plugins are automatically loaded when either of the applications is started.
```

Reviewer: Consider reordering to `Pupil Capture, Pupil Player, and Pupil Service`

> 

Reviewer 2: I think we need an additional paragraph regarding architectural best practices, specifically the logical split between the core algorithm and the PUpil plugin glue code.

Custom pupil detection plugins must subclass the [PupilDetectorPlugin class](https://github.com/pupil-labs/pupil/blob/master/pupil_src/shared_modules/pupil_detector_plugins/detector_base_plugin.py), which is available during runtime:

```py
from pupil_detector_plugins import PupilDetectorPlugin

class MyCustomPupilDetectorPlugin(PupilDetectorPlugin):

    def __init__(self, g_pool):
        super().__init__(g_pool)

>         # In most cases, it is desirable to disable other pupil detectors running in the eye process

Reviewer 2: ```suggestion
        # In some cases, it might be desirable to disable other pupil detectors running in the eye process
```

class MyCustomPlugin(Plugin):
    pass
```

### Pupil Detection Plugins

> Pupil Core supports custom pupil detection plugins that run in the eye process. These plugins can be used to implement custom algorithms for extracting pupillometry data from the eye images and feeding it to the rest of the gaze mapping pipeline.

Reviewer 2: ```suggestion
Starting with version v2.6, Pupil Core supports custom pupil detection plugins that run in the eye process. These plugins can be used to implement custom algorithms for extracting pupillometry data from the eye images and feeding it to the rest of the gaze mapping pipeline.
```

    def _stop_other_pupil_detectors(self):
        # Getting the Plugin_List instance from the g_pool object
        plugin_list = self.g_pool.plugins
        # Iterating over every plugin in the Plugin_List instance
        for plugin in plugin_list:
            # Deactivating every PupilDetectorPlugin instances except self
            if isinstance(plugin, PupilDetectorPlugin) and plugin is not self:
                plugin.alive = False
        # Forcing Plugin_List instance to remove deactivated plugins
        plugin_list.clean()

    @property
    def pupil_detector(self):
        # This read-only property must be implemented by the custom subclass.
        #
        # Returns an instance of pupil detector;
        # See: https://github.com/pupil-labs/pupil-detectors
        pass

    def detect(self, frame, **kwargs):
        # This method must be implemented by the custom subclass.
        #
        # Returns a pupil datum dictionary containing the pupil location and related metadata.
        # See Pupil Datum Format for a list of required keys:
        # https://docs.pupil-labs.com/developer/core/overview/#pupil-datum-format
        pass
```

> See an example of a custom pupil detection plugin which returns artificial pupil locations [here](https://gist.github.com/pupil-labs/todo-provide-uid-of-published-custom-detector-plugin).

Reviewer 2: You can find the public gist links here: https://github.com/pupil-labs/pupil-community/blob/master/README.md#pupil-detector-plugins
